src/context/component.js

```diff
@@ -0,0 +1,261 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {protectedNoInline, throttleTail} from './scheduler';
+import {pushIfNotExist, removeItem} from '../utils/array';
+import {withComponent} from './component-hooks';
+
+const EMPTY_ARRAY = [];
+const EMPTY_FUNC = () => {};
+
+/**
+ * @typedef {function(*, !./node.ContextNode, !Function, !Array<ContextProp>):!Component}
+ * @package
+ *
+ * The factory arguments are:
+ * 1. An opaque ID value.
+ * 2. The context node where component is attached.
+ * 3. The component's function.
+ * 4. An array of dependencies.
+ */
+export let ComponentFactoryDef;
+
+/**
+ * The component holder. It encapsulates the component's dependencies, input,
+ * internal state, and cleanup functions.
+ *
+ * @package
+ */
+export class Component {
+  /**
+   * @param {*} id An opaque component ID.
+   * @param {!./node.ContextNode} contextNode
+   * @param {!Function} func
+   * @param {!Array<!ContextProp>} deps
+   */
+  constructor(id, contextNode, func, deps) {
+    /** @package @const {*} */
+    this.id = id;
+
+    /** @package @const {!./node.ContextNode} */
+    this.contextNode = contextNode;
+
+    /** @private @const {!Function} */
+    this.func_ = func;
+
+    /** @private @const {!Array<!ContextProp>} */
+    this.deps_ = deps;
+
+    /** @private @const {!Array} */
+    this.depValues_ = deps.length > 0 ? deps.map(EMPTY_FUNC) : EMPTY_ARRAY;
+
+    /** @private @const {!Array<function(*)>} */
+    this.depSubscribers_ =
+      deps.length > 0
+        ? deps.map((dep, index) => (value) => {
+            this.depValues_[index] = value;
+            this.update_();
+          })
+        : EMPTY_ARRAY;
+
+    /** @private {*|undefined} */
+    this.input_ = undefined;
+
+    /** @private {boolean} */
+    this.running_ = false;
+
+    /** @private {?Array<{current: ?}>} */
+    this.refs_ = null;
+
+    /** @private {number} */
+    this.refPointer_ = -1;
+
+    /** @private {?function()} */
+    this.runCleanup_ = null;
+
+    /** @private {?Array<function()>} */
+    this.cleanups_ = null;
+
+    // Schedulers.
+    /** @private @const {function()} */
+    this.update_ = throttleTail(this.update_.bind(this), setTimeout);
```

Comment:
That's what we've done most of the time and we occasionally get bugs with someone calling `update_` instead of `throttleUpdate_`, which are really hard to spot in review and tests. So I decided to try something new within the context project :)